source/Files.cpp

```diff
@@ -148,352 +132,196 @@ void Files::Init(const char * const *argv)
+			list.emplace_back(entry);
 
-	closedir(dir);
-#endif
+	sort(list.begin(), list.end());
+	return list;
 }
 
 
 
-bool Files::Exists(const string &filePath)
+bool Files::Exists(const filesystem::path &filePath)
 {
-#if defined _WIN32
-	struct _stat buf;
-	return !_wstat(Utf8::ToUTF16(filePath).c_str(), &buf);
-#else
-	struct stat buf;
-	return !stat(filePath.c_str(), &buf);
-#endif
+	return exists(filePath);
 }
 
 
 
-time_t Files::Timestamp(const string &filePath)
+time_t Files::Timestamp(const filesystem::path &filePath)
 {
-#if defined _WIN32
-	struct _stat buf;
-	_wstat(Utf8::ToUTF16(filePath).c_str(), &buf);
-#else
-	struct stat buf;
-	stat(filePath.c_str(), &buf);
-#endif
-	return buf.st_mtime;
+	auto time = last_write_time(filePath);
+	auto systemTime = time_point_cast<chrono::system_clock::duration>(time - chrono::file_clock::now()
+			+ chrono::system_clock::now());
+	return chrono::system_clock::to_time_t(systemTime);
 }
 
 
 
-void Files::Copy(const string &from, const string &to)
+void Files::Copy(const filesystem::path &from, const filesystem::path &to)
 {
-#if defined _WIN32
-	CopyFileW(Utf8::ToUTF16(from).c_str(), Utf8::ToUTF16(to).c_str(), false);
-#else
-	Write(to, Read(from));
-	// Preserve the timestamps of the original file.
-	struct stat buf;
-	if(stat(from.c_str(), &buf))
-		Logger::LogError("Error: Cannot stat \"" + from + "\".");
-	else
-	{
-		struct utimbuf times;
-		times.actime = buf.st_atime;
-		times.modtime = buf.st_mtime;
-		if(utime(to.c_str(), &times))
-			Logger::LogError("Error: Failed to preserve the timestamps for \"" + to + "\".");
-	}
-#endif
+	copy(from, to, filesystem::copy_options::overwrite_existing);
 }
 
 
 
-void Files::Move(const string &from, const string &to)
+void Files::Move(const filesystem::path &from, const filesystem::path &to)
 {
-#if defined _WIN32
-	MoveFileExW(Utf8::ToUTF16(from).c_str(), Utf8::ToUTF16(to).c_str(), MOVEFILE_REPLACE_EXISTING);
-#else
-	rename(from.c_str(), to.c_str());
-#endif
+	rename(from, to);
 }
 
 
 
-void Files::Delete(const string &filePath)
+void Files::Delete(const filesystem::path &filePath)
 {
-#if defined _WIN32
-	DeleteFileW(Utf8::ToUTF16(filePath).c_str());
-#else
-	unlink(filePath.c_str());
-#endif
+	remove_all(filePath);
 }
 
 
 
 // Get the filename from a path.
-string Files::Name(const string &path)
+string Files::Name(const filesystem::path &path)
 {
-	// string::npos = -1, so if there is no '/' in the path this will
-	// return the entire string, i.e. path.substr(0).
-	return path.substr(path.rfind('/') + 1);
+	return path.filename().string();
 }
 
 
 
-FILE *Files::Open(const string &path, bool write)
+FILE *Files::Open(const filesystem::path &path, bool write)
 {
 #if defined _WIN32
 	FILE *file = nullptr;
-	_wfopen_s(&file, Utf8::ToUTF16(path).c_str(), write ? L"w" : L"rb");
+	_wfopen_s(&file, Utf8::ToUTF16(path.string()).c_str(), write ? L"w" : L"rb");
 	return file;
 #else
 	return fopen(path.c_str(), write ? "wb" : "rb");
```

Comment:
Why do we even have a separation between Windows and other OS's?

Reply:
Because C++ only supports opening files via streams, so we have to use the C-style api for the old FILE *.

It's something that will get refactored later.

Reply:
> Because C++ only supports opening files via streams, so we have to use the C-style api for the old FILE *.
> 
> It's something that will get refactored later.

As far as I'm aware, Windows allows you to use utf-8 in fopen, so the conversion/utf-16/_wfopen_s code is unnecessary.

Reply:
Interesting.

Well, this is the old code, so I don't see a reason to touch it. Though it is interesting that the default utf-8 path string conversions are deleted on windows.

Reply:
Maybe I'm slightly off (https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/fopen-wfopen?view=msvc-170#unicode-support), but with all the other paths operations losing their windows variants I see no point in keeping them for the Open functions.

Reply:
I replaced it with fopen; I don't know if it works.

It will be changed again when we switch to streams anyway.